### PR TITLE
Behavioural test fixes

### PR DESCRIPTION
**Pull request by glenna-m**
Changes to support the latest behavioural tests.  This includes support for stale data handling.

### CODE REVIEW

File: test/jsupdate-behaviour-mocha.js [hunk 1]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                    test.onOutPortData(network.processes.sut.component, 'output', adv);
+                    test.sendData(network.processes.once.component, 'input', "once");
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
+                    test.sendData(network.processes.once.component, 'input', "again");
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello once");
+    });
+    it("should produce output, but not fire updater if upstream updater did nothing, after error, after valid output", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js]: s/produce output/send an event/

File: test/jsupdate-behaviour-mocha.js [hunk 2]
```diff
@@ -13,10 +14,60 @@ var jswrapper = require('../src/javascript-wrapper');
  * Theses test are intented to verify the state management of the system.
  * @see https://github.com/rdf-pipeline/noflo-rdf-pipeline/issues/35
  */
-describe("framework-state", function() {
+describe("jsupdater-behaviour", function() {
+    it("should produce data from updater", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:18]: good

File: test/jsupdate-behaviour-mocha.js [hunk 3]
```diff
@@ -13,10 +14,60 @@ var jswrapper = require('../src/javascript-wrapper');
...
+                    return "Hello " + input;
+                })
+            }).then(function(network){
+                test.onOutPortData(network.processes.sut.component, 'error', fail);
+                test.onOutPortData(network.processes.sut.component, 'output', done);
+                test.sendData(network.processes.sut.component, 'input', "input");
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello input");
+    });
+    it("should produce promise result from updater", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:31]: good

File: test/jsupdate-behaviour-mocha.js [hunk 4]
```diff
@@ -13,10 +14,60 @@ var jswrapper = require('../src/javascript-wrapper');
...
+                    return Promise.resolve("Hello " + input);
+                })
+            }).then(function(network){
+                test.onOutPortData(network.processes.sut.component, 'error', fail);
+                test.onOutPortData(network.processes.sut.component, 'output', done);
+                test.sendData(network.processes.sut.component, 'input', "input");
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello input");
+    });
+    it("should produce an error when updater throws an error", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:44]: good

File: test/jsupdate-behaviour-mocha.js [hunk 5]
```diff
@@ -13,10 +14,60 @@ var jswrapper = require('../src/javascript-wrapper');
...
+                sut: jswrapper(function(input) {
+                    throw "Hello " + input;
+                })
+            }).then(function(network){
+                test.onOutPortData(network.processes.sut.component, 'error', done);
+                test.sendData(network.processes.sut.component, 'input', "input");
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello input");
+    });
+    it("should produce an error when updater rejects result", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:56]: good

File: test/jsupdate-behaviour-mocha.js [hunk 6]
```diff
@@ -29,7 +80,32 @@ describe("framework-state", function() {
                 test.onOutPortData(network.processes.sut.component, 'output', done);
                 test.sendData(network.processes.node1.component, 'in', "from node1");
                 test.sendData(network.processes.node2.component, 'in', "from node2");
-            });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello from node1 and from node2");
+    });
+    it("should fire updater when it has valid input states on all of its attached sockets", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:86]: good

File: test/jsupdate-behaviour-mocha.js [hunk 7]
```diff
@@ -61,14 +137,42 @@ describe("framework-state", function() {
                 network.graph.addEdge('sink', 'output', 'sut', 'input');
                 test.sendData(network.processes.sink.component, 'input', "tongue");
                 setTimeout(done.bind(this, "nothing happened"), 100);
-            });
+            }).catch(fail);
+        }).should.become("nothing happened");
+    });
+    it("should not fire updater if upstream updater returns undefined after producing an output state", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:143]: good

File: test/jsupdate-behaviour-mocha.js [hunk 8]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
                 test.sendData(network.processes.broken.component, 'input', "from broken");
                 test.onOutPortData(network.processes.sut.component, 'output', done);
                 test.onOutPortData(network.processes.sut.component, 'error', fail);
-            });
+            }).catch(fail);
         }).should.eventually.have.property('data', "Hello from broken");
     });
-    it("should not notify attached error port if an updater explicity sets the same error", function() {
+    it("should fire updater if upstream updater changed output state", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:224]: good

File: test/jsupdate-behaviour-mocha.js [hunk 9]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                    test.onOutPortData(network.processes.sut.component, 'output', adv);
+                    test.sendData(network.processes.once.component, 'input', "once");
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
+                    test.sendData(network.processes.once.component, 'input', "again");
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello again");
+    });
+    it("should fire updater if upstream updater changed output state lm", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:245]: good

File: test/jsupdate-behaviour-mocha.js [hunk 10]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                    test.onOutPortData(network.processes.sut.component, 'output', adv);
+                    test.sendData(network.processes.once.component, 'input', "once");
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
+                    test.sendData(network.processes.once.component, 'input', "again");
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello once");
+    });
+    it("should send an event, but not fire updater if upstream updater did nothing, after error, after valid output", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:270]: good

File: test/jsupdate-behaviour-mocha.js [hunk 11]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                        test.sendData(network.processes.upstream.component, 'input', "again");
+                    });
+                }).then(function() {
+                    test.onOutPortData(network.processes.sut.component, 'output', done);
+                    test.sendData(network.processes.upstream.component, 'input', "yet again");
+                });
+            }).catch(fail);
+        }).should.eventually.have.property('data', "Hello upstream");
+    });
+    it("should not fire updater if upstream updater corrected error, but has no state", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:308]: good

File: test/jsupdate-behaviour-mocha.js [hunk 12]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                    test.onOutPortData(network.processes.upstream.component, 'error', adv);
+                    test.sendData(network.processes.upstream.component, 'input', "upstream");
+                }).then(function() {
+                    test.sendData(network.processes.upstream.component, 'input', "correction");
+                    setTimeout(done.bind(this, "nothing happened"), 100);
+                });
+            }).catch(fail);
+        }).should.become("nothing happened");
+    });
+    it("should not fire updater if upstream updater set an error data", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:333]: s/set an error data/set and error data but returned undefined/

File: test/jsupdate-behaviour-mocha.js [hunk 13]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                    fail("Hello " + input);
+                })
+            }).then(function(network){
+                network.graph.addEdge('broken', 'output', 'sut', 'input');
+                test.sendData(network.processes.broken.component, 'input', "brake");
+                setTimeout(done.bind(this, "nothing happened"), 100);
+            });
+        }).should.become("nothing happened");
+    });
+    it("should fire error updater if upstream updater set an error data", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:353]: good

File: test/jsupdate-behaviour-mocha.js [hunk 14]
```diff
@@ -112,16 +218,162 @@ describe("framework-state", function() {
...
+                sut: jswrapper(function(input) {
+                    done("Hello " + input);
+                })
+            }).then(function(network){
+                network.graph.addEdge('broken', 'error', 'sut', 'input');
+                test.sendData(network.processes.broken.component, 'input', "from broken");
+            }).catch(fail);
+        }).should.become("Hello from broken");
+    });
+    it("should not notify attached error port if an updater explicity sets the same error", function() {
```
Comment (dbooth-boston) [inline test/jsupdate-behaviour-mocha.js:368]: good

**Comment by dbooth-boston** [comment]
The mocha tests look good.  I did not figure out the logic in framework-ondata, but if the behavioral tests work then let's go ahead with this, since hopefully the logic will get revamped at some point anyway, per issue #35.

File: src/javascript-wrapper.js [hunk 15]
```diff
@@ -17,45 +16,71 @@ var util = require('util');
...
+         var results = updater.apply(vni, updaterParameters);
+
+         if (! _.isUndefined(results)) { 
+             // Got some results back from updater
+             Promise.resolve(results).then(function(results) { 
+                 // If the updater returned anything, set the output state with it
+                 var newState = vni.outputState();
+                 if (newState.lm === oldOutputStateLm) { 
+                     // updater did not modify output state lm - create a new output state with new lm
+                     vni.outputState(createState(vni.vnid, results));
```
Comment (catch-point) [inline src/javascript-wrapper.js:69]: What should happen if the promise resolves to undefined? Is that valid data for the output state?

File: src/javascript-wrapper.js [hunk 16]
```diff
@@ -17,45 +16,71 @@ var util = require('util');
...
-             // If the updater returned anything, set the output state with it
-             vni.outputState( createState( vni.vnid, results ) );
+    // Execute the updater on the VNI context, passing the updater Parameters as the API arguments
+    return new Promise(function(resolve) { 
+
+         var results = updater.apply(vni, updaterParameters);
+
+         if (! _.isUndefined(results)) { 
+             // Got some results back from updater
+             Promise.resolve(results).then(function(results) {
```
Comment (catch-point) [inline src/javascript-wrapper.js]: This is forking the promise and this branch may not be executed before the framework-ondata checks the output state. Instead you should not fork the promise and include this as part of the primary promise flow.

**Comment by dbooth-boston** [comment]
LGTM

**Comment by catch-point** [comment]
LGTM

**Pull request merged**